linker/js/src/main/scala/org/scalajs/linker/irio/NodeVirtualIRFiles.scala

```diff
@@ -19,77 +19,245 @@ import scala.concurrent._
+      }
+    }.map(_.flatten)
+  }
+
+  def fromJar(path: String)(implicit ec: ExecutionContext): Future[NodeScalaJSIRContainer] =
+    stat(path).map(s => new NodeVirtualJarScalaJSIRContainer(path, version(s)))
+
+  def fromSingleFile(path: String, relativePath: String)(
+      implicit ec: ExecutionContext): Future[NodeScalaJSIRContainer] = {
+    stat(path).map(s => new NodeVirtualScalaJSIRFile(path, relativePath, version(s)))
+  }
 
-  def entryPointsInfo(implicit ec: ExecutionContext): Future[ir.EntryPointsInfo] =
-    withInputStream(ir.Serializers.deserializeEntryPointsInfo)
+  private def fromDirectory(dir: String)(
+      implicit ec: ExecutionContext): Future[Seq[NodeScalaJSIRContainer]] = {
+    cbFuture[js.Array[FS.Dirent]](FS.readdir(dir, ReadDirOpt, _)).flatMap { entries =>
+      val (dirs, files) = entries.toSeq.partition(_.isDirectory)
 
-  def tree(implicit ec: ExecutionContext): Future[ir.Trees.ClassDef] =
-    withInputStream(ir.Serializers.deserialize)
+      val subdirFiles = Future.traverse(dirs) { e =>
+        val path = Path.join(dir, e.name)
+        fromDirectory(path)
+      }
 
-  @inline
-  private def withInputStream[A](f: InputStream => A)(
-      implicit ec: ExecutionContext): Future[A] = {
-    def read() = {
-      val buf = new Uint8Array(NodeFS.readFileSync(path)).buffer
-      val stream = new ArrayBufferInputStream(buf)
-      try f(stream)
-      finally stream.close()
+      // Since we will remove relativePath (#3580) we do not bother calculating it here.
+      val irFileNames = files.map(_.name).filter(_.endsWith(".sjsir"))
+      val directFiles = Future.traverse(irFileNames) { name =>
+        val path = Path.join(dir, name)
+        fromSingleFile(path, "dummy")
+      }
+
+      for {
+        sdf <- subdirFiles
+        df <- directFiles
+      } yield sdf.flatten ++ df
     }
+  }
+
+  private def stat(path: String)(implicit ec: ExecutionContext): Future[FS.Stats] =
+    cbFuture[FS.Stats](FS.stat(path, _))
+
+  private def version(stats: FS.Stats): Option[String] =
+    stats.mtime.map(_.getTime.toString).toOption
 
-    VirtualScalaJSIRFile.withPathExceptionContext(path, Future(blocking(read())))
+  private def isNotFound(e: js.Error): Boolean = {
+    val code = e.asInstanceOf[js.Dynamic].code.asInstanceOf[js.UndefOr[String]]
+    code.exists(_ == "ENOENT")
   }
 }
 
-private[scalajs] class NodeVirtualJarScalaJSIRContainer(val path: String) extends ScalaJSIRContainer {
-  import NodeVirtualJarScalaJSIRContainer.JSZip
+private class NodeVirtualScalaJSIRFile(
+    val path: String,
+    val relativePath: String,
+    val version: Option[String]
+) extends VirtualScalaJSIRFile with NodeScalaJSIRContainer {
+  import NodeInterop._
 
-  val version: Option[String] =
-    NodeFS.statSync(path).mtime.map(_.getTime.toString).toOption
+  def entryPointsInfo(implicit ec: ExecutionContext): Future[ir.EntryPointsInfo] = {
+    def loop(fd: Int, buf: ByteBuffer): Future[ir.EntryPointsInfo] = {
```

Comment:
I think this loop never terminates if the file itself does not contain enough data for the entry points to be read. There should be a stop when `bytesRead` is 0 or some other condition triggered by Node.js' `fs.read` when the EOF has been reached.

Reply:
Euh... yes, indeed. Of course the Node.js documentation doesn't mention anything about that :(